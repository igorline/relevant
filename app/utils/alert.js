import { toast } from 'react-toastify';

const IS_NATIVE = process.env.WEB !== 'true';
const IS_BROWSER = process.env.BROWSER;

export const browserAlerts = {
  alert: (message, alertType) => {
    toast(message, {
      position: toast.POSITION.BOTTOM_RIGHT,
      type: alertType || 'error',
      hideProgressBar: true
    });
  }
};

export function Alert() {
  if (IS_NATIVE) return require('react-native').Alert;
  if (IS_BROWSER) return browserAlerts;

  // eslint-disable-next-line
  return { alert: (a, b) => console.log(a, ' ', b) };
}
