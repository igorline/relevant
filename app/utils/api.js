import * as tokenUtil from './token';

let post;
let routes = {};
let community;

// let postApi = '';
// let userApi = '';

console.log('BROWSER ', process.env.BROWSER);
console.log('WEB ', process.env.WEB);

if (process.env.BROWSER || process.env.WEB !== 'true') {
  // this is a weird hack that makes conditional require work in react-native
  // routes.post = require(postApi);
  // routes.user = require(userApi);
} else {
  // Desktop ONLY!!!
  // the if statment doesn't work anymore - user reat-native field in package.json
  // prevent react native from loading these modules
  let postApi = '../../server/api/post/post.controller';
  let userApi = '../../server/api/user/user.controller';
  // post = require(postApi);
  routes.post = require(postApi) || {};
  routes.user = require(userApi) || {};
}

const queryParams = (params) => {
  if (!params) return '';
  let paramString = Object.keys(params)
  .filter(p => params[p])
  .map(k => encodeURIComponent(k) + '=' + encodeURIComponent(params[k]))
  .join('&');
  if (paramString && paramString.length) return '?' + paramString;
  return '';
};


export function setCommunity(_community) {
  community = _community;
}


export function env() {
  if (process.env.WEB !== 'true') {
    require('../publicenv');
    return process.env;
  }
  return process.env;
}

export function Alert() {
  if (process.env.WEB !== 'true') {
    let ReactNative = require('react-native');
    let Platform = ReactNative.Platform;
    if (Platform.OS === 'ios') {
      return ReactNative.AlertIOS;
    }
    return ReactNative.Alert;
  } else if (process.env.BROWSER) {
    return window;
  }
  return { alert: (a, b) => console.log(a, ' ', b) };
}

/**
 * send request to api
 * @param  {[type]} options
 * query - Object of url query params
 * params - url params
 * endpoint - api endpoint
 * uri - optional - custom url
 * method - REST method
 * body: body
 */
export async function request(options) {
  let query = queryParams({ ...options.query, community });
  let apiPath = '/api/';
  if (options.endpoint.match('auth')) apiPath = '';
  let uri = options.uri || process.env.API_SERVER + apiPath + options.endpoint;
  let path = options.path || '';
  uri += path;

  try {
    if (options.params) {
      Object.keys(options.params).forEach(key => {
        uri += '/' + options.params[key];
      });
    }

    let response;
    let responseJSON;

    // ---------------------------------------------
    // This is the case when request is orginating from nodejs
    // ---------------------------------------------


    if (!process.env.BROWSER && process.env.WEB === 'true') {
      if (options.path === '' && options.params) options.path = 'findById';
      let req = {
        params: options.params,
        body: options.body,
        query: options.query,
      };
      let next = () => null;
      let res = null;
      responseJSON = await routes[options.endpoint][options.path](req, res, next);
      // in case we get a mongoose object back
      if (responseJSON && responseJSON.toObject) responseJSON = responseJSON.toObject();

    // ---------------------------------------------
    // This is the case when request is orginating from client
    // ---------------------------------------------
    } else {
      response = await fetch(uri + query, {
        method: options.method,
        ...await exports.reqOptions(),
        body: options.body
      });
      response = await exports.handleErrors(response);
      responseJSON = await response.json();
    }
    // throw new Error('omg');
    return responseJSON;
  } catch (error) {
    console.log('api error', uri, error);
    throw error;
  }
}

export async function reqOptions() {
  let token;
  try {
    token = await tokenUtil.get();
  } catch (err) {
    console.log('no token');
  }
  return {
    credentials: 'include',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    }
  };
}

export async function handleErrors(response) {
  if (!response.ok) {
    let error = response.statusText;
    try {
      let json = await response.json();
      if (json) {
        error = json.message;
        throw Error(error);
      }
    } catch (err) {
      throw Error(error);
    }
  }
  return response;
}
