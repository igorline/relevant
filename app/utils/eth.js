import crypto from 'crypto';

export async function signMessage(provider, address) {
  const signer = provider.getSigner();

  const now = new Date();
  const exp = Math.floor(now.setMinutes(now.getMinutes() + 5) / 1000);
  const msg = {
    address,
    exp
  };
  const signature = await signer.signMessage(JSON.stringify(msg));
  return { msg, signature };
}

export function truncateAddress(address) {
  if (!address) return null;
  return address.slice(0, 6) + '...' + address.slice(address.length - 4, address.length);
}

export function parseBN(value) {
  return value && value.get ? getBN(value) : value;
}

export function getBN(value) {
  const hex = value.get('_hex');
  if (hex === '0x00') {
    return 0;
  }
  return formatBN(hex, 18);
}

export function formatBN(hex, decimals = 18) {
  const numString = Number(hex).toString();
  // console.log('numString', numString);
  if (numString[numString.length - 3] === '+') {
    const trailingZeroes = getTrailingZeros(numString, decimals);
    const result = Number(withoutZeros(numString) + trailingZeroes);

    return Number.parseFloat(result).toFixed(0);
  }
  if (numString.length > 18) {
    return Number(numString.slice(0, -18));
  }
  if (numString.length === 18) {
    return Number(`0.${numString}`);
  }
  return Number(numString);
}

export function generateSalt() {
  return crypto.randomBytes(16).toString('hex');
}

export function formatBalanceRead(balString) {
  return `${balString.slice(0, -18)}.${balString.slice(-18)}`;
}

export function formatBalanceWrite(balString, decimals = 18) {
  return (balString * 10 ** decimals).toString();
}

export function appendZeroes(numString, amount) {
  return `${numString}${'0'.repeat(amount)}`;
}

export function removeDecimal(balString) {
  return balString.split('.').join('');
}

export function withoutZeros(numString) {
  return numString.slice(0, -2).toString();
}

export function getTrailingZeros(numString, decimals) {
  return Number(Number(numString.slice(-2)) - decimals).toString();
}
