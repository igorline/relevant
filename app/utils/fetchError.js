export function handleErrors(response) {
  console.log(response, 'error catching any response');
    if (!response.ok) {
        throw Error(response.statusText);
    }
    return response;
}