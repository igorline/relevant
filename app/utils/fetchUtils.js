import * as tokenUtil from './token';

const queryParams = (params) => {
  if (!params) return '';
  let paramString = Object.keys(params)
    .filter(p => params[p])
    .map(k => encodeURIComponent(k) + '=' + encodeURIComponent(params[k]))
    .join('&');
  if (paramString && paramString.length) return '?' + paramString;
  return '';
};


export function env() {
  if (process.env.WEB != 'true') {
    require('../publicenv');
    return process.env;
  }
  return process.env;
}

export function Alert() {
  if (process.env.WEB != 'true') {
    let Platform = require('react-native').Platform;
    if (Platform.OS === 'ios') {
      return require('react-native').AlertIOS;
    }
    return require('react-native').Alert;
  } else if (process.env.BROWSER) {
    return window;
  }
  return { alert: (a, b) => console.log(a, ' ', b) };
}

/**
 * [superFetch description]
 * @param  {[type]} options
 * params - Object of url query params
 * endpoint - api endpoint
 * uri - optional - custom url
 * method - REST method
 * body: body
 */
export async function superFetch(options) {
  let params = queryParams(options.params);
  let uri = options.uri || process.env.API_SERVER + '/api/' + options.endpoint;
  let path = options.path || '';
  try {
    let response = await fetch(uri + path + params, {
      method: options.method,
      ...await exports.reqOptions(),
      body: options.body
    });
    response = await exports.handleErrors(response);
    let responseJSON = await response.json();

    return responseJSON;
  } catch (error) {
    console.log('fetch error', uri, error);
    throw error;
  }
}

export async function reqOptions() {
  let token;
  try {
    token = await tokenUtil.get();
  } catch (err) {
    console.log('no token');
  }
  return {
    credentials: 'include',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    }
  };
}

export async function handleErrors(response) {
  if (!response.ok) {
    let error = response.statusText;
    try {
      let json = await response.json();
      if (json) {
        error = json.message;
        throw Error(error);
      }
    } catch (err) {
      throw Error(error);
    }
  }
  return response;
}
