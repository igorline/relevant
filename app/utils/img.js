import ExifReader from 'exifreader';
import dataUriToBuffer from 'data-uri-to-buffer';

/* Image resize / EXIF processing functions */

export const MAX_IMAGE_DIMENSION = 1024;
export const MAX_FILE_SIZE = 1024 * 400;

export function loadFile(file) {
  return new Promise(resolve => {
    const reader = new FileReader();
    reader.onload = e => resolve(e.target.result);
    reader.readAsArrayBuffer(file);
  });
}

function getScale(width, height, viewportWidth, viewportHeight, fillViewport) {
  function fitHorizontal() {
    return viewportWidth / width;
  }
  function fitVertical() {
    return viewportHeight / height;
  }
  fillViewport = !!fillViewport;
  const landscape = width / height > viewportWidth / viewportHeight;
  if (landscape) {
    if (fillViewport) {
      return fitVertical();
    }
    if (width > viewportWidth) {
      return fitHorizontal();
    }
  } else {
    if (fillViewport) {
      return fitHorizontal();
    }
    if (height > viewportHeight) {
      return fitVertical();
    }
  }
  return 1;
}

function applyRotation(canvas, ctx, deg) {
  const radians = deg * (Math.PI / 180);
  if (deg === 90) {
    ctx.translate(canvas.width, 0);
  } else if (deg === 180) {
    ctx.translate(canvas.width, canvas.height);
  } else if (deg === 270) {
    ctx.translate(0, canvas.height);
  }
  ctx.rotate(radians);
}

function base64ToUint8Array(string, start, finish) {
  start = start || 0;
  finish = finish || string.length;
  const binary = atob(string);
  const buffer = new Uint8Array(binary.length);
  for (let i = start; i < finish; i++) {
    buffer[i] = binary.charCodeAt(i);
  }
  return buffer;
}

function getOrientation(uri) {
  // Split off the base64 URL header
  const base64String = uri.split(',')[1];
  // Read off first 128KB, which is all we need to
  // get the EXIF data
  const arr = base64ToUint8Array(base64String, 0, 2 ** 17);
  try {
    const tags = ExifReader.load(arr.buffer);
    return tags.Orientation.value;
  } catch (err) {
    return 1;
  }
}

/**
 * Mapping from EXIF orientation values to data
 * regarding the rotation and mirroring necessary to
 * render the canvas correctly
 * Derived from:
 * http://www.daveperrett.com/articles/2012/07/28/exif-orientation-handling-is-a-ghetto/
 */
const orientationToTransform = {
  1: { rotation: 0, mirror: false },
  2: { rotation: 0, mirror: true },
  3: { rotation: 180, mirror: false },
  4: { rotation: 180, mirror: true },
  5: { rotation: 90, mirror: true },
  6: { rotation: 90, mirror: false },
  7: { rotation: 270, mirror: true },
  8: { rotation: 270, mirror: false }
};

function applyOrientationCorrection(canvas, ctx, uri) {
  const orientation = getOrientation(uri);
  // Only apply transform if there is some non-normal orientation
  if (orientation && orientation !== 1) {
    const { rotation } = orientationToTransform[orientation];
    const flipAspect = rotation === 90 || rotation === 270;
    if (flipAspect) {
      // Fancy schmancy swap algo
      canvas.width = canvas.height + canvas.width;
      canvas.height = canvas.width - canvas.height;
      canvas.width -= canvas.height;
    }
    if (rotation > 0) {
      applyRotation(canvas, ctx, rotation);
    }
  }
}

export function renderToCanvas(img, options) {
  if (!img) return null;
  options = options || {};

  // Canvas max size for any side
  const canvas = document.createElement('canvas');
  const ctx = canvas.getContext('2d');
  const initialScale = options.scale || 1;
  // Scale to needed to constrain canvas to max size
  let scale = getScale(
    img.width * initialScale,
    img.height * initialScale,
    MAX_IMAGE_DIMENSION,
    MAX_IMAGE_DIMENSION,
    true
  );
  // Still need to apply the user defined scale
  scale *= initialScale;
  canvas.width = Math.round(img.width * scale);
  canvas.height = Math.round(img.height * scale);
  const { correctOrientation } = options;
  const jpeg = !!img.src.match(/data:image\/jpeg|\.jpeg$|\.jpg$/i);
  const hasDataURI = !!img.src.match(/^data:/);

  ctx.save();

  // Can only correct orientation on JPEGs represented as dataURIs
  // for the time being
  if (correctOrientation && jpeg && hasDataURI) {
    applyOrientationCorrection(canvas, ctx, img.src);
  }
  // Resize image if too large
  if (scale !== 1) {
    ctx.scale(scale, scale);
  }

  ctx.drawImage(img, 0, 0);
  ctx.restore();

  return canvas;
}

export function downsample(img) {
  const resized = renderToCanvas(img, { correctOrientation: true });
  let buf;
  let dataURL;
  let quality = 0.8;
  do {
    dataURL = resized.toDataURL('image/jpeg', quality);
    buf = dataUriToBuffer(dataURL);
    quality -= 0.05;
  } while (buf.length > MAX_FILE_SIZE && quality > 0.1);
  return dataURL;
}

export function loadImage(file) {
  return new Promise((resolve, reject) => {
    if (file) {
      const reader = new FileReader();
      reader.onload = e => {
        reader.onload = null;
        reader.onerror = null;
        const dataURL = e.target.result;
        const buf = dataUriToBuffer(dataURL);
        if (buf.length <= MAX_FILE_SIZE) {
          resolve(dataURL);
        } else {
          const img = new Image();
          img.onload = () => {
            img.onload = null;
            img.onerror = null;
            resolve(downsample(img));
          };
          img.onerror = () => {
            img.onload = null;
            img.onerror = null;
            reject();
          };
          img.src = dataURL;
        }
      };
      reader.onerror = () => {
        reader.onload = null;
        reader.onerror = null;
        reject();
      };
      reader.readAsDataURL(file);
    } else reject();
  });
}
