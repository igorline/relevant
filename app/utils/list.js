export function unique(array) {
  return array.filter((elem, pos, arr) => arr.indexOf(elem) === pos);
}
