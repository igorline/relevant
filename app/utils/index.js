import * as s3 from './s3';
import * as post from './post';
import * as api from './api';
import * as storage from './storage';
import * as numbers from './numbers';
import * as text from './text';
import * as img from './img';
import * as alert from './alert';
import * as list from './list';
import * as tween from './tween';

let nav = {}; // eslint-disable-line

if (process.env.WEB !== 'true') {
  nav = require('./nav').default;
}

export { s3, post, api, alert, list, storage, numbers, text, nav, img, tween };
