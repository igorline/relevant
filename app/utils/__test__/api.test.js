import { _request } from 'utils/api';
import fetchMock from 'fetch-mock';

fetchMock.post('*', { hello: 'world' });

const getStore = () => ({
  community: {
    active: 'relevant'
  }
});

const options = {
  query: { search: 'user' },
  params: { _id: '123' },
  endpoint: 'user',
  method: 'POST',
  body: { data: 'some data' }
};

describe('client request', () => {
  it('generate correct fetch request', async () => {
    await _request(options, getStore);
    expect(fetchMock.lastCall()).toMatchSnapshot();
  });
});
