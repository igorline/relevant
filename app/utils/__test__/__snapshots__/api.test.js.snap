// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client request generate correct fetch request 1`] = `
Array [
  "/api/user/123?community=relevant&search=user",
  Object {
    "body": Object {
      "data": "some data",
    },
    "credentials": "include",
    "headers": Object {
      "Accept": "application/json",
      "Authorization": "Bearer undefined",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;
